### DISCUSSION TITLE
What are :param and :reader, etc officially called?

### DISCUSSION THREAD

**Discussion by yuki-kimoto** [Ovid/Cor#68] [Q&A] [answered]
What are :param and :reader etc officially called?

Attribute?

https://github.com/Ovid/Cor/blob/master/rfc/attributes.md#623-field-attributes

Modifier?

https://github.com/Ovid/Cor#kim

```
KEYWORD IDENTIFIER MODIFIER? SETUP?
```

If the name "Modifier" was used, the name is duplicated with "Method Modifier".

**Comment by Ovid** [comment]
They don't have an "official" name, though they should. [When I argued for the KIM (Keyword, Identifier, Modifier) syntax](https://ovid.github.io/articles/language-design-consistency.html), based on Damian Conway's feedback, I chose the term "modifier" largely to avoid the overloaded word "attribute." I was surprised at how easily people accepted the proposal, but we've struggled so long to even get the syntax and semantics down that we never formally agreed on naming.

This should probably be addressed because [[the `field` grammar uses the term "modifier" completely differently](https://github.com/Ovid/Cor/blob/master/rfc/grammar.md#33-field-grammar).

**Comment by haarg** [comment, marked as answer]
While the term "attribute" is a bit overloaded, in perl syntax it does have a specific meaning. And these modifiers are using that attribute syntax. So I'd say that these are attributes.

**Reply by Ovid** [reply]
Oops. I just renamed everything to modifier in the RFC. I'll revert that.

**Reply by Ovid** [reply]
Done.

**Comment by yuki-kimoto** [comment]
OK. I understand it is called ATTRIBUTE for now.

```
FIELD            ::= INSTANCE | SHARED ';'
SHARED           ::= 'my' { SCALAR | ARRAY | HASH } DEFAULT?
INSTANCE         ::= 'field'    FIELD_DEFINITION
FIELD_DEFINITION ::= SCALAR ATTRIBUTES? DEFAULT?  | { ARRAY | HASH } DEFAULT? 
DEFAULT          ::= '{' PERL_EXPRESSION '}'
ATTRIBUTE        ::= ':' ( 'param' MODIFIER? | 'reader' MODIFIER? | 'writer' MODIFIER? |  'predicate' MODIFIER?  | HANDLES )
ATTRIBUTES       ::= { ATTRIBUTE }
HANDLES          ::= 'handles' '(' 
                                    IDENTIFIER { ',' IDENTIFIER }    # list of methods this field handles
                                 |  PAIR { ',' PAIR }                # map of methods (to, from) this field handles
                                 | '*'                               # this field handles all unknown methods, but inheritance takes precedence
                              ')'
PAIR             ::= IDENTIFIER  ':' IDENTIFIER
MODIFIER         ::= '(' IDENTIFIER ')'
IDENTIFIER       ::= [:alpha:] {[:alnum:]}
```
